core/utils/poly-look/src/visualisations/charts/bubble-charts/bubbleCluster/bubbleCluster.js

```diff
@@ -165,9 +174,124 @@ export class BubbleCluster extends Chart {
       .attr("fill", this._textColor)
-      .attr("transform", (d) =>
-        d.data.icon ? `translate(${d.x},${d.y})` : ""
-      );
+      .attr("transform", (d) => (d.data.icon ? `translate(${d.x},${d.y})` : ""))
+      .on("click", this._onBubbleClick);
+  }
+
+  _addLabel(bubbleGroups) {
+    const label = this._label;
+    const color = this._textColor;
+    const chart = this.chart;
+    bubbleGroups.each(function (node) {
+      if (label(node)) {
+        const labelGroup = d3
+          .select(this)
+          .append("g")
+          .attr("class", "label-group");
+        const text = labelGroup
+          .append("text")
+          .attr("x", 0)
+          .attr("y", (d) => -d.r - labelMargin)
+          .attr("text-anchor", "middle")
+          .attr("fill", this._textColor)
+          .text(label);
+
+        text.each((_, i, nodes) => {
+          const bBox = nodes[i].getBBox();
+          labelGroup
+            .append("rect")
+            .attr("x", bBox.x - nodeLabelBoxPaddingX)
+            .attr("y", bBox.y - nodeLabelBoxPaddingY)
+            .attr("width", bBox.width + nodeLabelBoxPaddingX * 2)
+            .attr("height", bBox.height + nodeLabelBoxPaddingY * 2)
+            .attr("fill", defaultNodeLabelBoxColor)
+            .attr("opacity", defaultNodeLabelBoxOpacity)
+            .attr("rx", nodeLabelBoxBorderRadius);
+        });
+
+        const rect = labelGroup.select("rect");
+
+        const line = labelGroup
+          .append("line")
+          .attr("x1", 0)
+          .attr("x2", 0)
+          .attr("y1", (d) => -d.r - 2)
+          .attr("y2", (d) => -d.r - labelMargin + 7)
+          .attr("stroke", color)
+          .attr("stroke-width", 2);
+
+        const chartClientRect = chart.node().getBoundingClientRect();
+        const rectClientRect = rect.node().getBoundingClientRect();
+
+        if (chartClientRect.top == rectClientRect.top) {
+          text.attr("y", (d) => d.r + labelMargin + 10);
+
+          text.each((_, i, nodes) => {
+            const bBox = nodes[i].getBBox();
+            rect.attr("y", bBox.y - nodeLabelBoxPaddingY);
+          });
+
+          line
+            .attr("x1", 0)
+            .attr("x2", 0)
+            .attr("y1", (d) => d.r + 2)
+            .attr("y2", (d) => d.r + labelMargin - 7);
+        }
+
+        if (
+          Math.floor(chartClientRect.right) == Math.floor(rectClientRect.right)
+        ) {
+          text
+            .attr("x", (d) => -d.r - labelMargin)
+            .attr("y", 4)
+            .attr("text-anchor", "end");
+
+          text.each((_, i, nodes) => {
+            const bBox = nodes[i].getBBox();
+            rect
+              .attr("x", bBox.x - nodeLabelBoxPaddingX)
+              .attr("y", bBox.y - nodeLabelBoxPaddingY);
+          });
+
+          line
+            .attr("y1", 0)
+            .attr("y2", 0)
+            .attr("x1", (d) => -d.r - 2)
+            .attr("x2", (d) => -d.r - labelMargin + 7);
+        }
+
+        if (
```

Comment:
Would it be possible to un-DRY this code? Seems to be the same as above, only the mirror image.